### ISSUE TITLE
Restoring last read position occasionally fails, opening EPUB to beginning of chapter instead

### ISSUE THREAD

**Issue by ettore** [readium/r2-navigator-swift#153]
## Overview 

Reopening an EPUB trying to restore the last read position seem to work inconsistently, sometimes opening the book to the beginning of the chapter. This bug is currently blocking our transition to R2 which otherwise is complete. 

I am unable to reproduce the issue on iOS 14. However I can reproduce it roughly one time out of 5 attempts using the same steps, book, build and device on iOS 13, 12 and 10. (I never tried on 11.) It is easier to reproduce with long books trying to restore a position toward the end of one of the last chapters, but it's pretty random.

I can't reproduce this in the r2-testapp but I'll try to explain what I tried in detail what I am observing.

## Build / Test Set-up

- Xcode 11.5 toolchain
- carthage 0.36.0
- R2 version 2.0.0-beta.1
- test devices: iPhone 11 iOS 14.4 / iPhone 6 iOS 12.5.1
- simulators: iPads / iPhones iOS 13.5, 12.4, 10.3.1
- Focusing on epubs without DRM (e.g. Treasure Island, which is also present in the testapp). DRM'ed books exhibit same behavior though.

## R2 Set-up

The way R2 is integrated in our app (SimplyE) matches the r2-testapp's -- as a matter of fact I copied its architecture. As far as restoring the reading position is concerned, my understanding is that one simply has to pass the desired Locator to the `initialLocation` parameter in the `EPUBNavigatorViewController` initializer. Is this correct or is there more to it? 

I also verified the locator i am passing is always the same, with or without experiencing the bug.

## Investigation 

I've tried to follow how the `initialLocation` locator is propagated at runtime inside the R2 navigator code, to try to understand where it is getting lost. I see that it is passed to the [`reloadSpreads(at:)` function](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L370), and there i verified a non-zero `initialIndex` is always found.

The code then flows to [`PaginationView::setCurrentIndex(_:location:)`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/Toolkit/PaginationView.swift#L186) and then [`loadView(at:location:completion:)`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/Toolkit/PaginationView.swift#L228). I verified the guards in those functions are not triggered.

From there control passes to [`EPUBReflowableSpreadView::go(to:completion:)`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBReflowableSpreadView.swift#L195), which may be called multiple times during initialization from many places. I assume this is expected? 

This is where things get confusing. When the bug happens the last `go(to:completion:)` call is for a Locator with `progression == 0`. This seems to [come from a polling timer](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L331), for updating the user setting style. In this last function invocation we [call `go(to:)` with the `currentLocation` value](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L346), which is incorrect because its `progression` is `0`. Shouldn't that still be the `initialLocation` i passed in in the navigator's initializer, since the reader is still setting itself up? (There's a spinner on the screen at this time.) I just can't figure out where that `initialLocation` ends up going, or why/if it's being overwritten.

From another POV, I noticed that on the r2-testapp `navigator(_:locationDidChange:)` is called twice (once with `progression==0`, then with the correct locator) but on our app it's only called _once_, and when the bug happens the `locator.progression` is `0`. That callback is also called from [another polling timer](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L439), which makes me think order of execution is not guaranteed to be identical across 2 identical instantiations of the navigator.

I also noticed that if i start putting a lot of breakpoints with debugger output (which slows down execution), the bug is more likely to be triggered. This hints at possible race conditions, but it's just a conjecture.

## Long write-up, sorry... final thoughts

I've tried many other approaches (e.g. trying to follow the spreads set-up, or how the Locator `progression` property is being set) but probably i should stop here since I'm rambling at this point.

Thanks in advance for any help/pointers/suggestions. I realize the fact that the r2-testapp is fine probably means I have a bug in my set-up, but on the other hand this random behavior could mean something is off inside r2-navigator, but I ran out of things to look into.

**Comment by mickael-menu** [comment]
Thanks for your detailed write-up.

> As far as restoring the reading position is concerned, my understanding is that one simply has to pass the desired Locator to the `initialLocation` parameter in the `EPUBNavigatorViewController` initializer. Is this correct or is there more to it?

Correct.

> From there control passes to [`EPUBReflowableSpreadView::go(to:completion:)`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBReflowableSpreadView.swift#L195), which may be called multiple times during initialization from many places. I assume this is expected?

It depends, this is called for example when applying the styles as well, so we might get several calls during initial load. Besides, by default we are preloading several resources which might muddy your logs. I suggest initializing `EPUBNavigatorViewController` with this configuration:

```swift
var config = EPUBNavigatorController.Configuration()
config.preloadPreviousPositionCount = 0
config.preloadNextPositionCount = 0
config.debugState = true
```

`config.debugState = true` will print out all the state changes in the navigator in the logs. That might help see what is happening.

> This is where things get confusing. When the bug happens the last `go(to:completion:)` call is for a Locator with `progression == 0`. This seems to [come from a polling timer](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L331), for updating the user setting style. In this last function invocation we [call `go(to:)` with the `currentLocation` value](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L346), which is incorrect because its `progression` is `0`. Shouldn't that still be the `initialLocation` i passed in in the navigator's initializer, since the reader is still setting itself up?

Yes, actually the `initialLocation` is saved in the [`.jumping(pendingLocator)`](https://github.com/readium/r2-navigator-swift/blob/f9c268da74b3ae8f387af3dffbd1589b4641b16e/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L91) state. The [`currentLocation` property](https://github.com/readium/r2-navigator-swift/blob/f9c268da74b3ae8f387af3dffbd1589b4641b16e/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L401)  is dynamically computed from the `pendingLocator` if the state is `.jumping`, otherwise it will use the current scroll progression in the web view.

Thanks to the polling timer, `updateUserSettingStyle()` is called only when the state is `.idle`, to avoid overriding the pending locator (what you observed). So my guess is that in your situation, [we transition to `.idle`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L111) before the web view actually finished laying out and jumping to the pending locator.


[The crux of the issue is probably here](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/Toolkit/PaginationView.swift#L302-L307). Unfortunately, I could not find any way to get a proper event from the web view to know when the layout/positioning is finished, so I reverted to a naive delay which might fail with large resources. This is reminiscent of [this Kotlin issue](https://github.com/readium/r2-navigator-kotlin/issues/177).

I would be very happy to review any contribution for this issue, if someone has a eureka moment. I guess we could hack something in JavaScript to check that we actually scrolled in the web view to more or less the expected progression. Until then you might "fix" the issue by increasing the delay duration.

Also I think there's a more fundamental issue in letting the reading app request to update user styles explicitly after initializing the navigator. If the styles timing was entirely in the control of the navigator, we could make sure everything (including styles) is properly loaded in the web view before attempting to load the initial location. But such changes would probably take place in [`Navigator2`](https://github.com/readium/r2-testapp-kotlin/discussions/394#discussion-3241296).


> From another POV, I noticed that on the r2-testapp `navigator(_:locationDidChange:)` is called twice (once with `progression==0`, then with the correct locator) but on our app it's only called _once_, and when the bug happens the `locator.progression` is `0`.

Are you sure you tested with the latest version of `r2-testapp-swift`? I can see only a single `locationDidChange` event, we skip the one with the `0` progression because it's actually invalid (still pending). That also means that the bug you reported might be in the test app too.

**Comment by ettore** [comment]
> ```swift
> var config = EPUBNavigatorController.Configuration()
> config.preloadPreviousPositionCount = 0
> config.preloadNextPositionCount = 0
> config.debugState = true
> ```

I tried this and the bug does seem harder to reproduce. I couldn't reproduce it anymore on an iPad simulator on iOS 13, but I still can on iOS 12.4 and 10.3.1.

> >we [call `go(to:)` with the `currentLocation` value](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L346), which is incorrect because its `progression` is `0`. Shouldn't that still be the `initialLocation` i passed in in the navigator's initializer, since the reader is still setting itself up?
> 
> Yes, actually the `initialLocation` is saved in the [`.jumping(pendingLocator)`](https://github.com/readium/r2-navigator-swift/blob/f9c268da74b3ae8f387af3dffbd1589b4641b16e/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L91) state. The [`currentLocation` property](https://github.com/readium/r2-navigator-swift/blob/f9c268da74b3ae8f387af3dffbd1589b4641b16e/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L401) is dynamically computed from the `pendingLocator` if the state is `.jumping`, otherwise it will use the current scroll progression in the web view.

Ah, very interesting. I do see [the transition to `.jumping`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L107), but the `if case let .jumping(pendingLocator) = state` in the [`currentLocation`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L404) computed property is never hit. Both in the testapp and in our app. So it seems like the `initialLocation` is obtained from somewhere else, which is probably problematic.

> Thanks to the polling timer, `updateUserSettingStyle()` is called only when the state is `.idle`, to avoid overriding the pending locator (what you observed). So my guess is that in your situation, [we transition to `.idle`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L111) before the web view actually finished laying out and jumping to the pending locator.

hmm can you point me in the code where we know when the web view has completed layout?

I mapped the transitions for success/fail cases across a few iOS versions, and it seems that when it fails `locationDidChange` is called too late, except on iOS 10 where it's called twice for some reason. (I omitted the extra "transition" from `loading` to `loading` at the very beginning which is just an initialization artifact.)

**iOS 13.5, 12.4 success**
loading
idle
locationDidChange(initialLocator)
jumping
idle

**iOS 12.4 fail**
loading
idle
jumping
idle
locationDidChange(progression = 0)

**iOS 10.3.1, success**
loading
idle
locationDidChange(initialLocator)
jumping
idle
locationDidChange(initialLocator)

**iOS 10.3.1, fail**
loading
idle
locationDidChange(progression == 0)
jumping (progression == 0)
idle
locationDidChange(progression == 0)

> [The crux of the issue is probably here](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/Toolkit/PaginationView.swift#L302-L307). Unfortunately, I could not find any way to get a proper event from the web view to know when the layout/positioning is finished, so I reverted to a naive delay which might fail with large resources. This is reminiscent of [this Kotlin issue](https://github.com/readium/r2-navigator-kotlin/issues/177).

This would also kind of explain why I am seeing the problem on older devices / OS versions: with less computing power, that layout phase might just get slow enough to screw things up.

Increasing that delay does _seem_ to make the bug harder to reproduce. I was wondering if the [EPUBReflowableSpreadView::spreadDidLoad() delay](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBReflowableSpreadView.swift#L148-L149) or [the one in `go(to:)`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBReflowableSpreadView.swift#L181-L183) might also be involved?

> I would be very happy to review any contribution for this issue, if someone has a eureka moment. I guess we could hack something in JavaScript to check that we actually scrolled in the web view to more or less the expected progression. 

I'd be very happy to -- my knowledge of JS is not up-to-par though. My one thought was to see if there's a way to synchronize things with a serial queue instead of custom delays, but I still have a way to go in terms of getting familiar with the r2navigator internals to understand if that's even possible. Is there documentation for me to read that goes over these details and workarounds?

> Until then you might "fix" the issue by increasing the delay duration.

Sorry, did you mean [the one in PaginationView](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/Toolkit/PaginationView.swift#L302-L307)? I also tried to increase the polling timer for [`notifyCurrentLocation`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L442) and [`_updateUserSettingsStyle`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L333) update but that made not much difference.
 
> Are you sure you tested with the latest version of `r2-testapp-swift`? I can see only a single `locationDidChange` event, we skip the one with the `0` progression because it's actually invalid (still pending). That also means that the bug you reported might be in the test app too.

Well, I'm using the [2.2.0-beta.1 tag](https://github.com/readium/r2-testapp-swift/tree/2.2.0-beta.1) because it is bound to the last beta release of R2 (`2.0.0-beta.1`), and for our app that seemed more stable than just using the top of `develop`.  `2.0.0-beta.1` seems pretty recent to me, but is it actually too old?

But yes you're right, `navigator(_:locationDidChange:)` is called only once except on iOS 10 where it's called twice.

Also .... this does look like an heisenbug in that it's made worse by the additional debug logging probably interacting with those custom delays.

**Comment by mickael-menu** [comment]
> I tried this and the bug does seem harder to reproduce. I couldn't reproduce it anymore on an iPad simulator on iOS 13, but I still can on iOS 12.4 and 10.3.1.

Ha yes, not so surprising in retrospect...

> Ah, very interesting. I do see [the transition to `.jumping`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L107), but the `if case let .jumping(pendingLocator) = state` in the [`currentLocation`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L404) computed property is never hit. Both in the testapp and in our app. So it seems like the `initialLocation` is obtained from somewhere else, which is probably problematic.

That should be normal actually, I don't think `currentLocation` is called until the publication is loaded. But the initial locator is also provided to `PaginationView.reloadAtIndex()`. The one stored in `.jumping` is only used in case the app request the `currentLocator` (e.g. when creating a bookmark) before the navigator settles to the pending locator.
  
> hmm can you point me in the code where we know when the web view has completed layout?

We don't, that's the issue :/ We just rely on this delay before assuming the page is loaded: https://github.com/readium/r2-navigator-swift/blob/f9c268da74b3ae8f387af3dffbd1589b4641b16e/r2-navigator-swift/Toolkit/PaginationView.swift#L299-L308

> I'd be very happy to -- my knowledge of JS is not up-to-par though. My one thought was to see if there's a way to synchronize things with a serial queue instead of custom delays, but I still have a way to go in terms of getting familiar with the r2navigator internals to understand if that's even possible. Is there documentation for me to read that goes over these details and workarounds?

Unfortunately we don't have any in-depth documentation at the moment.

 
> Sorry, did you mean [the one in PaginationView](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/Toolkit/PaginationView.swift#L302-L307)? I also tried to increase the polling timer for [`notifyCurrentLocation`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L442) and [`_updateUserSettingsStyle`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L333) update but that made not much difference.

I believe you should increase the one from [`PaginationView`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/Toolkit/PaginationView.swift#L302-L307)

* The one in [`EPUBReflowableSpreadView.spreadDidLoad()`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBReflowableSpreadView.swift#L148-L149) is used to wait for the web view pagination to settles before attempting to scroll to the target progression, otherwise we might end up at the wrong spot.
* [This one](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/EPUB/EPUBReflowableSpreadView.swift#L181-L183) is only used when turning pages in a single resource. If the page turn is animated, there's a delay roughly the length of the animation before notifying the page turn completion. This should not have any impact on the issue you reported.
 
But as you can see with the comments I put in the code, I'm not very satisfied with these delays. It's a crutch and very fragile in edge cases. [I think in R1 we were also relying on delays in `readium-shared-js`](https://github.com/readium/readium-shared-js/blob/7f245beba1ed97eaabce0aa5e9cf2f3b23e8f8f6/js/views/reader_view.js#L301-L306), but maybe there's a better way.

> Well, I'm using the [2.2.0-beta.1 tag](https://github.com/readium/r2-testapp-swift/tree/2.2.0-beta.1) because it is bound to the last beta release of R2 (`2.0.0-beta.1`), and for our app that seemed more stable than just using the top of `develop`. `2.0.0-beta.1` seems pretty recent to me, but is it actually too old?
> But yes you're right, `navigator(_:locationDidChange:)` is called only once except on iOS 10 where it's called twice.

No it should be fine actually, I'm not sure why you see two locator events on iOS 10.

**Comment by ettore** [comment]
> I believe you should increase the one from [`PaginationView`](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/Toolkit/PaginationView.swift#L302-L307)

Upon further investigation, by increasing this delay to something like 2 seconds i noticed that when the bug happens, the webview is _already_ displaying the correct page (!) -- but then `PaginationView::goToIndex(_:location:animated:completion)` is called with the wrong locator (progression == 0). So at that point actually increasing the delay doesn't change anything.

However, I think I have found a related bug: I noticed that in that `goToIndex(...)` function the ["to delay or not to delay" check](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/Toolkit/PaginationView.swift#L304) was failing even for locators where `progression` was not 0. I fixed that in this WIP PR https://github.com/readium/r2-navigator-swift/pull/154 also integrating also some of your comments which I think are useful.

With that fix, and your navigator initialization configuration suggestion, things improved quite a bit: the bug happens half the time in iOS 12 and 11 i.e. roughly once every 12 times vs 1 out of 6. However, iOS 10 is still horrible (1 out of 3).

I tried one more thing. Since the "bad" locator originated from the polling call on [`_updateUserSettingsStyle`](https://github.com/readium/r2-navigator-swift/blob/f3c779b3464373c6c5428aab58cfb8bd4ffcbadb/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L346), I tried to increase that polling timer from 0.1s to 1.0s. This seems to improve things dramatically for iOS 10 to the point that I no longer see the bug on iOS 10 (!) -- I have not had time to re-check the other iOS's but I will. However, I don't know if that's ok to increase or if it has negative side effects elsewhere.

**Comment by ettore** [comment]
Sadly I have to take my last comment back. I confirm that by increasing the [`_updateUserSettingsStyle` polling time](https://github.com/readium/r2-navigator-swift/blob/f3c779b3464373c6c5428aab58cfb8bd4ffcbadb/r2-navigator-swift/EPUB/EPUBNavigatorViewController.swift#L333) to 1 second things get pretty reliable on the simulators on all iOS versions, but when i tried on a physical iPhone 6 on iOS 12.5 the bug shows itself pretty often (say roughly once every 5 attempts). I wonder if it's because my old iPhone 6 is super slow in general, but increasing the polling interval to something much larger brings no difference, unlike on the simulators.

Increasing the [PaginationView delay](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/Toolkit/PaginationView.swift#L305) also makes no difference (on the other hand this is identical to the simulator). 

Back to square one... not sure what else i can tweak... :(

**Comment by mickael-menu** [comment]
> However, I think I have found a related bug: I noticed that in that `goToIndex(...)` function the ["to delay or not to delay" check](https://github.com/readium/r2-navigator-swift/blob/db2ed2b5f4c60e1f2ddb5c8e30795ecd9998a7cc/r2-navigator-swift/Toolkit/PaginationView.swift#L304) was failing even for locators where `progression` was not 0. I fixed that in this WIP PR #154 also integrating also some of your comments which I think are useful.

Well spotted, thanks!
 
> Back to square one... not sure what else i can tweak... :(

Nothing comes to mind either without debugging myself... Did you try to reproduce it on the test app from your iPhone 6?

**Comment by ettore** [comment]
> Nothing comes to mind either without debugging myself... Did you try to reproduce it on the test app from your iPhone 6?

I did, but it doesn't reproduce there. I do see the beginning of the chapter "flashing" before jumping to the correct position, but that's ok. My remaining hypothesis/guess is that our app perhaps spends some cycles doing some other tasks (network calls, etc) and that is enough to mess with the delays in some way. I will try to profile runs between the 2 apps. 

I also tried to compile the app with release optimizations, but that also brings no difference.

On the other hand, if I had to create a sample app to try to show this bug, I would write something like the r2-testapp, but that... does not show the bug.

**Comment by ettore** [comment]
A sliver of hope.... it looks like my build system broke in some way when I started building for device, and I was building the app with the unmodified version of r2-navigator-swift. Once I fixed that my iPhone 6 / iOS 12 started behaving like the simulators! I'm going to ask a few people to try my "fix" on other devices / other OS versions and if all goes well I can add this to my PR.

**Comment by ettore** [comment]
Alright, I updated my PR with my latest findings. I also noted all the QA that we have done, since that's very important given the aleatoric nature of this bug. 

I realize this is not a fix, but it's a patch that doesn't change the current behavior of the navigator, which is something I am not ready to tackle at this point. 

For that, I started thinking that maybe the state machine needs to be expanded to consider additional states. It's actually really nice to have that in the navigator. For example I saw that sometimes R2 flashes the correct page on the screen, then reverts back to the beginning of the chapter with a `jumping(pendingLocator: progression==0)` transition, which is clearly wrong. (NB: with the expanded polling interval that transition no longer happens, of course.) I think if we could define the fact that we are still initializing the Navigator, then we could just ignore that 2nd transition, maybe, without resorting to hard-coded delays. But I have not given it in-depth thought, and there's too much of the navigator code that I do not know.

But for the time being (and given the urgency my team has) that patch hopefully is ok.

**Comment by mickael-menu** [comment]
Thanks for carrying this through, I'll review your PR this week!

> I think if we could define the fact that we are still initializing the Navigator, then we could just ignore that 2nd transition, maybe, without resorting to hard-coded delays. But I have not given it in-depth thought, and there's too much of the navigator code that I do not know.

Technically that's already what the `loading` state is for. But as long as we don't have a proper "laid out" event coming from the JS layer, we are stuck with delays. In any case, if we find an alternative to the delays, the state machine should already handle this properly.

**Comment by ettore** [comment]
Thanks @mickael-menu for merging the PR. Do you think there will soon be a new beta (or even alpha) tag pushed across all the R2 repos to include those fixes in an official release?

**Comment by mickael-menu** [comment]
Probably in the next few weeks.
